presto-open-telemetry/src/main/java/com/facebook/presto/opentelemetry/OpenTelemetryTracer.java

```diff
@@ -0,0 +1,203 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.opentelemetry;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.tracing.Tracer;
+import com.facebook.presto.tracing.TracingConfig;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.baggage.Baggage;
+import io.opentelemetry.api.baggage.propagation.W3CBaggagePropagator;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.facebook.presto.spi.StandardErrorCode.DISTRIBUTED_TRACING_ERROR;
+
+public class OpenTelemetryTracer
+        implements Tracer
+{
+    private static String currentContextPropagator = TracingConfig.ContextPropagator.B3_SINGLE_HEADER;
+    private static OpenTelemetry openTelemetry = OpenTelemetryBuilder.build(currentContextPropagator);
+    private final io.opentelemetry.api.trace.Tracer openTelemetryTracer;
+    private final String traceToken;
+    private final Span parentSpan;
+    private final Context parentContext;
+
+    public final Map<String, Span> spanMap = new ConcurrentHashMap<>();
+    public final Map<String, Span> recorderSpanMap = new LinkedHashMap<>();
+
+    public OpenTelemetryTracer(String traceToken, String contextPropagator, String propagatedContext, String baggage)
+    {
+        // Trivial getter method to return carrier
+        // Carrier will be equal to the data to be fetched
+        TextMapGetter<String> trivialGetter = new TextMapGetter<String>()
+        {
+            @Override
+            public String get(String carrier, String key)
+            {
+                return carrier;
+            }
+
+            @Override
+            public Iterable<String> keys(String carrier)
+            {
+                return Arrays.asList(get(carrier, null));
+            }
+        };
+
+        // Rebuild OPEN_TELEMETRY instance if necessary (to use different context propagator)
+        // Will only occur once at max, if contextPropagator is different from B3_SINGLE_HEADER
+        if (contextPropagator != null && !contextPropagator.equals(currentContextPropagator)) {
+            this.openTelemetry = OpenTelemetryBuilder.build(contextPropagator);
+            this.currentContextPropagator = contextPropagator;
```

Comment:
I'm not expert in open telemetry but are we sure the logic is correct by modifying **static** variables?

Reply:
I believe it should be fine. If we have a different context propagator, we need to rebuild the open telemetry instance for all tracers that may exist.